invoices/invoiceregistry.go

```diff
@@ -1043,3 +955,114 @@ func (i *InvoiceRegistry) HodlUnsubscribeAll(subscriber chan<- interface{}) {
 
 	delete(i.hodlReverseSubscriptions, subscriber)
 }
+
+// invoiceUpdater is an object that is responsible for updating the invoice
+// state and storing actions that need to be carried out after the invoice is
+// updated.
+type invoiceUpdater struct {
+	// update input fields
+	circuitKey           channeldb.CircuitKey
+	amtPaid              lnwire.MilliSatoshi
+	expiry               uint32
+	currentHeight        int32
+	finalCltvRejectDelta int32
+
+	// update result fields
+	updateSubscribers bool
+	result            string
+}
+
+// update is a callback for DB.UpdateInvoice that contains the invoice
+// settlement logic.
+func (i *invoiceUpdater) update(inv *channeldb.Invoice) (
+	*channeldb.InvoiceUpdateDesc, error) {
+
+	// Don't update the invoice when this is a replayed htlc.
+	htlc, ok := inv.Htlcs[i.circuitKey]
+	if ok {
+		switch htlc.State {
+		case channeldb.HtlcStateCanceled:
+			i.result = "replayed htlc to canceled invoice"
```

Comment:
this is a bit awkward. Could we just debug log here instead?

Reply:
Then I need to pass in another parameter (the payment hash), to keep the log line the same. I think it is no problem to leave it as is. @carlaKC is working on exposing these failure reasons and this also prepares for that.

Reply:
Could we go straight to an enum for `result`? I'd have to string match or replace with an enum as is. 
I'll also need `NotifyExitHopHtlc` to return the outcome of the update, but that's less relevant for this PR.

Reply:
Enum created